### ISSUE TITLE
[Feature] Complete DeepSeek V4 MTP prefill/decode execution contract

### ISSUE THREAD

**Issue by hashiqiqixian** [hw-native-sys/pypto-lib#678]
### Summary

Complete the DeepSeek V4 MTP execution contract in `pypto-lib`, covering both kernel-side MTP computation and serving-facing prefill/decode state construction.

This feature should make the remaining MTP work reviewable as one coherent feature:

1. define the exact DeepSeek V4 MTP projection shape contract;
2. define the MTP decoder-layer and logits boundaries;
3. define prefill/decode input construction helpers;
4. define speculative-token state updates;
5. provide one independently runnable full-chain validation case.

### Motivation / Use Case

DeepSeek V4 MTP requires more than a standalone projection kernel. The full path needs a clear contract for how target-model hidden states, MTP input tokens, decoder-layer execution, logits computation, and speculative-token state are connected.

The current gap is that individual pieces can be implemented or validated separately, but the repository still needs a complete, reviewable MTP execution contract that explains:

1. what hidden state should be passed into MTP;
2. how MTP projection handles HC lanes;
3. where the MTP forward boundary ends;
4. where logits are computed;
5. how prefill and decode construct MTP inputs;
6. how speculative tokens are verified and updated.

Without this contract, it is easy to accidentally use post-`hc_head` hidden states, treat lane-aware MTP projection as a normal `[T, D]` projection, or mix forward and logits behavior into one unclear boundary.

### Proposed API / Behavior

Add the remaining DeepSeek V4 MTP implementation and validation pieces under `models/deepseek/v4`.

Required files:

- `models/deepseek/v4/mtp.py`
- `models/deepseek/v4/mtp_inputs.py`

Required MTP kernel/validation functions:

- `mtp_projection_impl`
- `mtp_seed_hc_stack`
- `mtp_decoder_layer_tail`
- `mtp_forward_tail`
- `mtp_compute_logits_tail`
- `mtp_full_chain`

Required serving-facing helpers:

- `build_mtp_prefill_input_ids`
- `restore_cp_prefill_next_tokens`
- `build_mtp_prefill_input_ids_cp`
- `build_main_decode_input_ids_for_mtp`
- `build_mtp_decode_input_ids`
- `verify_mtp_spec_tokens`
- `update_mtp_state_after_step`

Expected MTP projection behavior:

```text
previous_hidden_states
    -> reshape to [T, HC_MULT, D]
    -> hnorm
    -> h_proj

input embedding / current hidden state
    -> mask positions == 0
    -> enorm
    -> e_proj
    -> unsqueeze/broadcast to HC lane dimension

projected_mtp_hidden = h_branch + e_branch
projected_mtp_hidden: [T, HC_MULT, D]
```

Expected MTP decoder-layer boundary:

```text
projected_mtp_hidden [T, HC_MULT, D]
    -> MHC pre / attention norm boundary
    -> SWA attention
    -> MHC post/pre / FFN norm boundary
    -> MoE/FFN
    -> MHC post
    -> pre_hc_residual [T, HC_MULT * D]
```

Expected logits boundary:

```text
pre_hc_residual
    -> hc_head
    -> shared_head_norm
    -> lm_head logits
```

MTP forward should return `pre_hc_residual [T, HC_MULT * D]`. Logits should be computed by a separate logits tail.

For multi-step MTP, the same step index must select the same MTP layer for forward and logits. Each proposal step should feed the previous step's flat pre-`hc_head` residual back as the next `previous_hidden_states`.

Only one runnable validation case is required:

```text
full-chain
```

The `full-chain` case should cover:

```text
prefill/decode input construction
    -> lane-aware projection
    -> decoder-layer tail
    -> logits tail
    -> candidate logits / speculative token state
```

### Alternatives Considered

One alternative is to only validate the projection kernel first and leave decoder tail, logits tail, and serving-state helpers for later. This makes the first change smaller, but it does not provide a complete MTP execution contract and makes review harder because later code has to infer the missing boundaries.

Another alternative is to put all MTP behavior directly into serving code. That would make the runtime path less reusable and harder to validate independently in `pypto-lib`.

The proposed approach keeps kernel-side MTP computation and CPU-validatable serving-state helpers separated, while still providing one full-chain validation case to verify the complete path.

### Additional Context

Important model contract details:

- MTP must consume the target model's pre-`hc_head` flat residual, not post-`hc_head` or final normalized hidden states.
- `previous_hidden_states` should support `[T, HC_MULT * D]` or `[T, HC_MULT, D]`.
- MTP projection is lane-aware and should produce `[T, HC_MULT, D]`.
- The `positions == 0` embedding mask is part of the MTP input contract.
- MTP forward and MTP logits should remain separate boundaries.
- The decode proposal loop must recycle the previous step's flat pre-`hc_head` residual.
- Partitioned prefill token restore and packed input construction should be covered.
- Only one full-chain validation case is required for this issue.

**Comment by high-cloud** [comment]
### CANN DeepSeek V4 MTP reference and current PyPTO gap

I re-checked the current `cann-recipes-infer/models/deepseek-v4` implementation and the latest local PyPTO-Lib DeepSeek V4 sweep notes. This is the reference behavior and the measured gap that this issue should track.

#### Reference implementation shape in `cann-recipes-infer`

- MTP is enabled by `model_config.next_n`.
  - A3 CI config uses `next_n: 1`.
  - 950 CI config uses `next_n: 3`.
  - `infer.py` rejects `next_n > 3`.
- `infer.py` constructs a separate MTP runner when `next_n > 0` and then shares the main model's `embed_tokens`, `lm_head`, and `rotary_emb` with the MTP runner.
- `runner_deepseek.py` initializes `DeepseekV3ModelMTP` for the MTP runner.
- `modeling_deepseek.py` implements the MTP model as a special one-layer tail model:
  - `DeepseekV3ModelMTPLayer` starts at `config.num_hidden_layers`.
  - it builds `num_nextn_predict_layers` MTP decoder layers under prefix `mtp.0`;
  - it unsqueezes/repeats hidden states over HC lanes, runs the MTP decoder layer, then applies `hc_head`;
  - the MTP model has no independent `lm_head`; the runner shares the main model's `lm_head`.
- Attention in MTP mode forces `compress_ratio = 1`, so the draft path is the C1A/SWA-style attention path rather than the compressed C4A/C128A path.
- `model_infer.py` owns the speculative loop:
  - main model prefill/decode samples `main_next_tokens`;
  - decode verifies MTP `spec_tokens` against `main_next_tokens[:, :next_n]` and produces per-batch `accepted_num`;
  - prefill switches both main and MTP input dicts into decode mode and seeds MTP with `main_hidden`;
  - decode advances `kv_len`, `input_lens`, `generate_ids`, `prev_hidden_states`, and MTP `kv_len_cached` according to `accepted_num`;
  - `mtp_post_process` runs the MTP model `next_n` times, greedily samples one draft token per step, accumulates `spec_tokens`, and feeds the latest `prev_hidden_states` into the next draft step.

The important implementation boundary is therefore not just `mtp_projection`. The feature needs all of these pieces to be connected: shared embeddings/head/rotary, MTP tail decoder, HC-head/logits boundary, accepted-token state update, MTP cache lengths, and the host-side draft-token loop.

#### Reference model measurement to keep in mind

The CANN DeepSeek V4 inference guide reports MTP-based benchmark numbers and acceptance-rate assumptions:

- 950DT Flash: 16-card deployment, 8K/128K benchmark sections; the guide notes MTP3 data with average `1.44` accepted tokens out of `3` draft tokens.
- Atlas A3 Flash: 64-card deployment, 8K decode section; the guide notes MTP1 data with average `0.7` accepted tokens out of `1` draft token.

So PyPTO-Lib needs to be able to report not only per-kernel latency, but also an equivalent-token latency or throughput measurement that accounts for accepted draft tokens.

#### Current PyPTO-Lib measured status / gap

From the local 2026-06-25 DeepSeek V4 NPU sweep notes:

- `models/deepseek/v4/mtp_projection.py` is the only current MTP-specific runnable kernel in the tracked tree.
  - It passes standalone real-NPU validation.
  - Measured latency: about `852.00us`.
  - Precision criterion used: `hidden_states_out: atol=1e-2, rtol=1e-2, max_error_ratio=0.02`.
- The current tracked `models/deepseek/v4` file list has `mtp_projection.py` but no complete `decode_mtp.py`, `prefill_mtp.py`, `mtp.py`, or full-chain MTP entrypoint.
- Main-model pieces are further along than MTP:
  - single-layer decode SWA/CSA/HCA passed two-card L2 validation;
  - single-layer prefill passed after retry/no-L2 coverage;
  - `decode_fwd.py` 43-layer two-card runtime completed with outer runtime around `80.41s`, but `golden_fn=None`, so this is runtime completion rather than golden precision validation;
  - `prefill_fwd.py --num-layers 1 --kernel-only` completed with outer runtime around `14.89s`, also without golden comparison.

Current gap relative to the CANN reference:

1. no full MTP runner/entrypoint that composes projection -> MTP decoder tail -> HC head -> shared head norm/lm head;
2. no serving-facing state builder for `spec_tokens`, `accepted_num`, MTP `kv_len`, `kv_len_cached`, `prev_hidden_states`, and main-model verification input ids;
3. no full-chain validation case that measures accepted-token behavior or equivalent latency;
4. no real-weight MTP validation against the Flash checkpoint path;
5. no acceptance-rate/performance report comparable to the CANN MTP1/MTP3 benchmark assumptions.

A good completion target for this issue is one runnable MTP full-chain case that can print both numerical validation results and MTP accounting (`draft tokens`, `accepted_num`, equivalent latency), even if the first version only targets MTP1 before generalizing to MTP3.
